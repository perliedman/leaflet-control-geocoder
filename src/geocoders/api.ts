import * as L from 'leaflet';

/**
 * An object that represents a result from a geocoding query
 */
export interface GeocodingResult {
  /**
   * Name of found location
   */
  name: string;
  /**
   * The bounds of the location
   */
  bbox: L.LatLngBounds;
  /**
   * The center coordinate of the location
   */
  center: L.LatLng;
  /**
   * URL for icon representing result; optional
   */
  icon?: string;
  /**
   * HTML formatted representation of the name
   */
  html?: string;
  /**
   * Additional properties returned by the geocoder
   */
  properties?: any;
}

/**
 * A callback function used in {@link IGeocoder.geocode} and {@link IGeocoder.suggest} and {@link IGeocoder.reverse}
 */
export type GeocodingCallback = (result: GeocodingResult[]) => void;

/**
 * An interface implemented to respond to geocoding queries
 */
export interface IGeocoder {
  /**
   * Performs a geocoding query and returns the results to the callback in the provided context
   * @param query the query
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  geocode(query: string, cb: GeocodingCallback, context?: any): void;
  /**
   * Performs a geocoding query suggestion (this happens while typing) and returns the results to the callback in the provided context
   * @param query the query
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  suggest?(query: string, cb: GeocodingCallback, context?: any): void;
  /**
   * Performs a reverse geocoding query and returns the results to the callback in the provided context
   * @param location the coordinate to reverse geocode
   * @param scale the map scale possibly used for reverse geocoding
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  reverse?(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void;
}

export interface GeocoderOptions {
  /**
   * URL of the service
   */
  serviceUrl: string;
  /**
   * Additional URL parameters (strings) that will be added to geocoding requests
   */
  geocodingQueryParams?: Record<string, unknown>;
  /**
   * Additional URL parameters (strings) that will be added to reverse geocoding requests
   */
  reverseQueryParams?: Record<string, unknown>;
  /**
   * API key to use this service
   */
  apiKey?: string;
}

/**
 * @internal
 */
export function geocodingParams(
  options: GeocoderOptions,
  params: Record<string, unknown>
): Record<string, unknown> {
  return L.Util.extend(params, options.geocodingQueryParams);
}

/**
 * @internal
 */
export function reverseParams(
  options: GeocoderOptions,
  params: Record<string, unknown>
): Record<string, unknown> {
  return L.Util.extend(params, options.reverseQueryParams);
}
