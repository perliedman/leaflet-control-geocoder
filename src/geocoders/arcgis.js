import L from 'leaflet';
import { getJSON } from '../util';

export default {
  class: L.Class.extend({
    options: {
      service_url: 'http://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer'
    },

    initialize: function(accessToken, options) {
      L.setOptions(this, options);
      this._accessToken = accessToken;
    },

    geocode: function(query, cb, context) {
      var params = {
        SingleLine: query,
        outFields: 'Addr_Type',
        forStorage: false,
        maxLocations: 10,
        f: 'json'
      };

      if (this._key && this._key.length) {
        params.token = this._key;
      }

      getJSON(this.options.service_url + '/findAddressCandidates', params, function(data) {
        var results = [],
          loc,
          latLng,
          latLngBounds;

        if (data.candidates && data.candidates.length) {
          for (var i = 0; i <= data.candidates.length - 1; i++) {
            loc = data.candidates[i];
            latLng = L.latLng(loc.location.y, loc.location.x);
            latLngBounds = L.latLngBounds(
              L.latLng(loc.extent.ymax, loc.extent.xmax),
              L.latLng(loc.extent.ymin, loc.extent.xmin)
            );
            results[i] = {
              name: loc.address,
              bbox: latLngBounds,
              center: latLng
            };
          }
        }

        cb.call(context, results);
      });
    },

    suggest: function(query, cb, context) {
      return this.geocode(query, cb, context);
    },

    reverse: function(location, scale, cb, context) {
      var params = {
        location: encodeURIComponent(location.lng) + ',' + encodeURIComponent(location.lat),
        distance: 100,
        f: 'json'
      };

      getJSON(this.options.service_url + '/reverseGeocode', params, function(data) {
        var result = [],
          loc;

        if (data && !data.error) {
          loc = L.latLng(data.location.y, data.location.x);
          result.push({
            name: data.address.Match_addr,
            center: loc,
            bounds: L.latLngBounds(loc, loc)
          });
        }

        cb.call(context, result);
      });
    }
  }),

  factory: function(accessToken, options) {
    return new L.Control.Geocoder.ArcGis(accessToken, options);
  }
};
