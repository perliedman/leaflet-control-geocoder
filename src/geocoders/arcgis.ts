import * as L from 'leaflet';
import { getJSON } from '../util';
import {
  IGeocoder,
  GeocoderOptions,
  GeocodingCallback,
  geocodingParams,
  GeocodingResult,
  reverseParams
} from './api';

export interface ArcGisOptions extends GeocoderOptions {}

/**
 * Implementation of the [ArcGIS geocoder](https://developers.arcgis.com/features/geocoding/)
 */
export class ArcGis implements IGeocoder {
  options: ArcGisOptions = {
    serviceUrl: 'https://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer',
    apiKey: ''
  };

  constructor(options?: Partial<ArcGisOptions>) {
    L.Util.setOptions(this, options);
  }

  geocode(query: string, cb: GeocodingCallback, context?: any): void {
    const params = geocodingParams(this.options, {
      token: this.options.apiKey,
      SingleLine: query,
      outFields: 'Addr_Type',
      forStorage: false,
      maxLocations: 10,
      f: 'json'
    });

    getJSON(this.options.serviceUrl + '/findAddressCandidates', params, data => {
      const results: GeocodingResult[] = [];
      if (data.candidates && data.candidates.length) {
        for (let i = 0; i <= data.candidates.length - 1; i++) {
          const loc = data.candidates[i];
          const latLng = L.latLng(loc.location.y, loc.location.x);
          const latLngBounds = L.latLngBounds(
            L.latLng(loc.extent.ymax, loc.extent.xmax),
            L.latLng(loc.extent.ymin, loc.extent.xmin)
          );
          results[i] = {
            name: loc.address,
            bbox: latLngBounds,
            center: latLng
          };
        }
      }

      cb.call(context, results);
    });
  }

  suggest(query: string, cb: GeocodingCallback, context?: any): void {
    return this.geocode(query, cb, context);
  }

  reverse(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void {
    const params = reverseParams(this.options, {
      location: encodeURIComponent(location.lng) + ',' + encodeURIComponent(location.lat),
      distance: 100,
      f: 'json'
    });
    getJSON(this.options.serviceUrl + '/reverseGeocode', params, data => {
      const result: GeocodingResult[] = [];
      if (data && !data.error) {
        const center = L.latLng(data.location.y, data.location.x);
        const bbox = L.latLngBounds(center, center);
        result.push({
          name: data.address.Match_addr,
          center: center,
          bbox: bbox
        });
      }

      cb.call(context, result);
    });
  }
}

/**
 * [Class factory method](https://leafletjs.com/reference.html#class-class-factories) for {@link ArcGis}
 * @param options the options
 */
export function arcgis(options?: Partial<ArcGisOptions>) {
  return new ArcGis(options);
}
