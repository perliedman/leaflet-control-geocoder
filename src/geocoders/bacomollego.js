var L = require('leaflet'),
	Util = require('../util'),
	proj4 = require('proj4');

module.exports = {
	class: L.Class.extend({
		options: {
			tmerc: new proj4.Proj("+proj=tmerc +lat_0=-34.629269 +lon_0=-58.463300 +k=0.999998 +x_0=100000 +y_0=100085 +ellps=intl +units=m +no_defs"),
			wsg: new proj4.Proj("EPSG:4326"),
			geocodeUrl: 'https://ws.usig.buenosaires.gob.ar/geocoder/2.2/geocoding/',
			reverseUrl: 'https://ws.usig.buenosaires.gob.ar/geocoder/2.2/reversegeocoding/',
			normalizarUrl: 'https://servicios.usig.buenosaires.gob.ar/normalizar',
			timeout: 10000,
			maxOptions: 3
		},

		initialize: function(options) {
			L.setOptions(this, options);
		},

		_normalizar: function(query, cb, context){
			Util.jsonp(this.options.normalizarUrl, {
				maxOptions: this.options.maxOptions,
				direccion: query
			}, function(json){
				cb.call(context, json);
			}, this, "callback");
			return this;
		},

		_tmerc2LatLng: function(a){
			var b = proj4.toPoint(a);
			return proj4.transform(this.options.tmerc, this.options.wsg, b),
			[ b.y, b.x ];
		},

		geocode: function(query, cb, context) {
			var results = [],
				maxOptions = 0,
				timedOut = false,
				timer;

			function checkIfCb(){
				if(!timedOut && results.length >= maxOptions){
					timedOut = true;
					cb.call(context || cb, results);
					clearTimeout(timer);
				}
			}

			function pushAndCb(c){
				results.push(c);
				checkIfCb();
			}

			this._normalizar(query, function(json){
				timer = setTimeout(L.bind(function(){
					timedOut = true;
					cb.call(context || cb, results);
				}, this), this.options.timeout);

				maxOptions = json.direccionesNormalizadas.length;

				json.direccionesNormalizadas.forEach(function(e){
					if(e.coordenadas){
						var center = L.latLng(e.coordenadas.y, e.coordenadas.x);
						pushAndCb({
							bbox: L.latLngBounds(center, center),
							name: e.direccion,
							center: center
						});
					}else if([
							(e.nombre_localidad||"").toUpperCase(),
							(e.nombre_partido||"").toUpperCase(),
					].indexOf("CABA") > -1)
						L.bind(function(e){
							Util.jsonp(this.options.geocodeUrl, {
								cod_calle: e.cod_calle,
								altura: e.altura
							}, function(tmerc){
								if(typeof(tmerc.x) != "undefined" && typeof(tmerc.y) != "undefined"){
									var center = L.latLng.apply(L, this._tmerc2LatLng([tmerc.x, tmerc.y]));
									pushAndCb({
										bbox: L.latLngBounds(center, center),
										name: e.direccion,
										center: center
									});
								}else{
									maxOptions--;
									checkIfCb();
								}
							}, this, "callback");
						}, this)(e);
					else
						this._normalizar(e.direccion, function(json){
							var data = json.direccionesNormalizadas[0];
							if(data && data.coordenadas){
								var center = L.latLng(data.coordenadas.y, data.coordenadas.x);
								pushAndCb({
									bbox: L.latLngBounds(center, center),
									name: data.direccion,
									center: center
								});
							} else{
								maxOptions--;
								checkIfCb();
							}
						}, this);
				}, this);
			}, this);
		},

		suggest: function(query, cb, context) {
			return this.geocode(query, cb, context);
		},

		reverse: function(location, scale, cb, context) {
			Util.jsonp(this.options.reverseUrl, {
				x: location.lng,
				y: location.lat,
			}, function(json){
				var results = [{
					bbox: L.latLngBounds(location, location),
					name: json.puerta,
					center: location
				}];
				cb.call(context || cb, results);
			}, "callback");
		}
	}),

	factory: function(options) {
		return new L.Control.Geocoder.BAComoLlego(options);
	}
};

