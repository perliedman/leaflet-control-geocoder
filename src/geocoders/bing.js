import L from 'leaflet';
import { jsonp } from '../util';

export default {
  class: L.Class.extend({
    initialize: function(key) {
      this.key = key;
    },

    geocode: function(query, cb, context) {
      jsonp(
        'https://dev.virtualearth.net/REST/v1/Locations',
        {
          query: query,
          key: this.key
        },
        function(data) {
          var results = [];
          if (data.resourceSets.length > 0) {
            for (var i = data.resourceSets[0].resources.length - 1; i >= 0; i--) {
              var resource = data.resourceSets[0].resources[i],
                bbox = resource.bbox;
              results[i] = {
                name: resource.name,
                bbox: L.latLngBounds([bbox[0], bbox[1]], [bbox[2], bbox[3]]),
                center: L.latLng(resource.point.coordinates)
              };
            }
          }
          cb.call(context, results);
        },
        this,
        'jsonp'
      );
    },

    reverse: function(location, scale, cb, context) {
      jsonp(
        '//dev.virtualearth.net/REST/v1/Locations/' + location.lat + ',' + location.lng,
        {
          key: this.key
        },
        function(data) {
          var results = [];
          for (var i = data.resourceSets[0].resources.length - 1; i >= 0; i--) {
            var resource = data.resourceSets[0].resources[i],
              bbox = resource.bbox;
            results[i] = {
              name: resource.name,
              bbox: L.latLngBounds([bbox[0], bbox[1]], [bbox[2], bbox[3]]),
              center: L.latLng(resource.point.coordinates)
            };
          }
          cb.call(context, results);
        },
        this,
        'jsonp'
      );
    }
  }),

  factory: function(key) {
    return new L.Control.Geocoder.Bing(key);
  }
};
