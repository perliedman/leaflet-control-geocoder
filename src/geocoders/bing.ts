import * as L from 'leaflet';
import { jsonp } from '../util';
import {
  IGeocoder,
  GeocoderOptions,
  GeocodingCallback,
  geocodingParams,
  GeocodingResult,
  reverseParams
} from './api';

export interface BingOptions extends GeocoderOptions {}

/**
 * Implementation of the [Bing Locations API](https://docs.microsoft.com/en-us/bingmaps/rest-services/locations/)
 */
export class Bing implements IGeocoder {
  options: BingOptions = {
    serviceUrl: 'https://dev.virtualearth.net/REST/v1/Locations'
  };

  constructor(options?: Partial<BingOptions>) {
    L.Util.setOptions(this, options);
  }

  geocode(query: string, cb: GeocodingCallback, context?: any): void {
    const params = geocodingParams(this.options, {
      query: query,
      key: this.options.apiKey
    });
    jsonp(
      this.options.apiKey,
      params,
      data => {
        const results: GeocodingResult[] = [];
        if (data.resourceSets.length > 0) {
          for (let i = data.resourceSets[0].resources.length - 1; i >= 0; i--) {
            const resource = data.resourceSets[0].resources[i],
              bbox = resource.bbox;
            results[i] = {
              name: resource.name,
              bbox: L.latLngBounds([bbox[0], bbox[1]], [bbox[2], bbox[3]]),
              center: L.latLng(resource.point.coordinates)
            };
          }
        }
        cb.call(context, results);
      },
      this,
      'jsonp'
    );
  }

  reverse(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void {
    const params = reverseParams(this.options, {
      key: this.options.apiKey
    });
    jsonp(
      this.options.serviceUrl + location.lat + ',' + location.lng,
      params,
      data => {
        const results: GeocodingResult[] = [];
        for (let i = data.resourceSets[0].resources.length - 1; i >= 0; i--) {
          const resource = data.resourceSets[0].resources[i],
            bbox = resource.bbox;
          results[i] = {
            name: resource.name,
            bbox: L.latLngBounds([bbox[0], bbox[1]], [bbox[2], bbox[3]]),
            center: L.latLng(resource.point.coordinates)
          };
        }
        cb.call(context, results);
      },
      this,
      'jsonp'
    );
  }
}

/**
 * [Class factory method](https://leafletjs.com/reference.html#class-class-factories) for {@link Bing}
 * @param options the options
 */
export function bing(options?: Partial<BingOptions>) {
  return new Bing(options);
}
