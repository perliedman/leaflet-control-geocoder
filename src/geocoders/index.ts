export * from './api';
export * from './arcgis';
export * from './bing';
export * from './google';
export * from './here';
export * from './latlng';
export * from './mapbox';
export * from './mapquest';
export * from './neutrino';
export * from './nominatim';
export * from './open-location-code';
export * from './opencage';
export * from './pelias';
export * from './photon';
export * from './what3words';
