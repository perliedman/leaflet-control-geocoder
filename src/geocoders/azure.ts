import * as L from 'leaflet';
import { getJSON } from '../util';
import { GeocodingResult, IGeocoder } from './api';

export interface AzureMapsOptions {
  apiKey: string; // Azure Maps API Key
  serviceUrl: string; // Optional: Base URL for the Azure Maps API
}

/**
 * Implementation of [Azure Maps Geocoding](https://www.microsoft.com/en-us/maps/azure/location-services/geocoding)
 *
 * https://learn.microsoft.com/en-us/rest/api/maps/search?view=rest-maps-1.0
 */
export class AzureMaps implements IGeocoder {
  private options: AzureMapsOptions = {
    apiKey: '',
    serviceUrl: 'https://atlas.microsoft.com/search'
  };

  constructor(options: Partial<AzureMapsOptions>) {
    L.Util.setOptions(this, options);
    if (!this.options.apiKey) {
      throw new Error('Azure Maps Geocoder requires an API key.');
    }
  }

  /**
   * {@inheritdoc}
   * https://learn.microsoft.com/en-us/rest/api/maps/search/get-search-address?view=rest-maps-1.0&tabs=HTTP
   */
  async geocode(query: string): Promise<GeocodingResult[]> {
    const params = {
      'api-version': '1.0',
      query,
      'subscription-key': this.options.apiKey
    };
    const url = this.options.serviceUrl + '/address/json';
    const data = await getJSON<AzureMapsResponse>(url, params);

    return (data.results || []).map(
      (result): GeocodingResult => ({
        name: result.address.freeformAddress,
        bbox: new L.LatLngBounds(
          [result.viewport.topLeftPoint.lat, result.viewport.topLeftPoint.lon],
          [result.viewport.btmRightPoint.lat, result.viewport.btmRightPoint.lon]
        ),
        center: new L.LatLng(result.position.lat, result.position.lon)
      })
    );
  }

  /**
   * {@inheritdoc}
   * https://learn.microsoft.com/en-us/rest/api/maps/search/get-search-address-reverse?view=rest-maps-1.0&tabs=HTTP
   */
  async reverse(location: L.LatLngLiteral, scale: number): Promise<GeocodingResult[]> {
    const params = {
      'api-version': '1.0',
      query: location.lat + ',' + location.lng,
      'subscription-key': this.options.apiKey
    };
    const url = this.options.serviceUrl + '/address/reverse/json';
    const data = await getJSON<any>(url, params);

    return (data.addresses || []).map(
      (address): GeocodingResult => ({
        name: address.address.freeformAddress,
        bbox: new L.LatLngBounds(
          [address.viewport.topLeftPoint.lat, address.viewport.topLeftPoint.lon],
          [address.viewport.btmRightPoint.lat, address.viewport.btmRightPoint.lon]
        ),
        center: new L.LatLng(location.lat, location.lng)
      })
    );
  }
}

/**
 * [Class factory method](https://leafletjs.com/reference.html#class-class-factories) for {@link Azure}
 * @param options the options
 */
export function azure(options: AzureMapsOptions) {
  return new AzureMaps(options);
}

/**
 * @internal
 */
export interface AzureMapsResponse {
  summary: Summary;
  results: Result[];
}

interface Result {
  type: string;
  id: string;
  score: number;
  address: Address;
  position: Position;
  viewport: Viewport;
  entryPoints: EntryPoint[];
}

interface Address {
  streetNumber: string;
  streetName: string;
  municipalitySubdivision: string;
  municipality: string;
  countrySecondarySubdivision: string;
  countryTertiarySubdivision: string;
  countrySubdivisionCode: string;
  postalCode: string;
  extendedPostalCode: string;
  countryCode: string;
  country: string;
  countryCodeISO3: string;
  freeformAddress: string;
  countrySubdivisionName: string;
}

interface EntryPoint {
  type: string;
  position: Position;
}

interface Position {
  lat: number;
  lon: number;
}

interface Viewport {
  topLeftPoint: Position;
  btmRightPoint: Position;
}

interface Summary {
  query: string;
  queryType: string;
  queryTime: number;
  numResults: number;
  offset: number;
  totalResults: number;
  fuzzyLevel: number;
}
