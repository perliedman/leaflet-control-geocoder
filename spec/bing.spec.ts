import { afterEach, describe, expect, it, vi } from 'vitest';
import { mockFetchRequest } from './mockFetchRequest';
import { Bing } from '../src/geocoders/bing';
import { GeocodingResult } from '../src/geocoders/api';

describe('L.Control.Geocoder.Bing', () => {
  afterEach(() => vi.clearAllMocks());
  it('geocodes Microsoft Way, Redmond', async () => {
    const geocoder = new Bing({ apiKey: '0123xyz' });
    const result = await mockFetchRequest(
      'https://dev.virtualearth.net/REST/v1/Locations/?query=Microsoft+Way%2C+Redmond&key=0123xyz',
      {
        authenticationResultCode: 'ValidCredentials',
        brandLogoUri: 'http:\/\/dev.virtualearth.net\/Branding\/logo_powered_by.png',
        copyright:
          'Copyright Â© 2011 Microsoft and its suppliers. All rights reserved. This API cannot be accessed and the content and any results may not be used, reproduced or transmitted in any manner without express written permission from Microsoft Corporation.',
        resourceSets: [
          {
            estimatedTotal: 1,
            resources: [
              {
                __type: 'Location:http:\/\/schemas.microsoft.com\/search\/local\/ws\/rest\/v1',
                bbox: [
                  47.636257744012461, -122.13735364288299, 47.643983179153814, -122.12206713944467
                ],
                name: '1 Microsoft Way, Redmond, WA 98052',
                point: {
                  type: 'Point',
                  coordinates: [47.640120461583138, -122.12971039116383]
                },
                address: {
                  addressLine: '1 Microsoft Way',
                  adminDistrict: 'WA',
                  adminDistrict2: 'King Co.',
                  countryRegion: 'United States',
                  formattedAddress: '1 Microsoft Way, Redmond, WA 98052',
                  locality: 'Redmond',
                  postalCode: '98052'
                },
                confidence: 'High',
                entityType: 'Address',
                geocodePoints: [
                  {
                    type: 'Point',
                    coordinates: [47.640120461583138, -122.12971039116383],
                    calculationMethod: 'InterpolationOffset',
                    usageTypes: ['Display']
                  },
                  {
                    type: 'Point',
                    coordinates: [47.640144601464272, -122.12976671755314],
                    calculationMethod: 'Interpolation',
                    usageTypes: ['Route']
                  }
                ],
                matchCodes: ['Good']
              }
            ]
          }
        ],
        statusCode: 200,
        statusDescription: 'OK',
        traceId: 'b0b1286504404eafa7e7dad3e749d570'
      },
      () => geocoder.geocode('Microsoft Way, Redmond')
    );

    const feature: GeocodingResult = result[0];
    expect(feature.name).toBe('1 Microsoft Way, Redmond, WA 98052');
    expect(feature.center).toStrictEqual({ lat: 47.64012046158314, lng: -122.12971039116383 });
    expect([[result]]).toMatchSnapshot();
  });
});
