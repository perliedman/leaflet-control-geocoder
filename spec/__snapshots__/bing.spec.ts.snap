// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`L.Control.Geocoder.Bing > geocodes Microsoft Way, Redmond 1`] = `
[
  [
    [
      {
        "bbox": {
          "_northEast": {
            "lat": 47.643983179153814,
            "lng": -122.12206713944467,
          },
          "_southWest": {
            "lat": 47.63625774401246,
            "lng": -122.13735364288299,
          },
        },
        "center": {
          "lat": 47.64012046158314,
          "lng": -122.12971039116383,
        },
        "name": "1 Microsoft Way, Redmond, WA 98052",
      },
    ],
  ],
]
`;
